### PR TITLE
PLT-4990 (WebApp): Separate channel management permissions.

### PR DESCRIPTION
**Pull request by grundleborg**
#### Summary

Separate channel management permissions for public/private create/manage/delete in System Console, and make sure the UI elements affected by these permissions elsewhere are shown/hidden in the correct situations.

Channel-Admin permission level will come later in a separate PR.

#### Ticket Link

https://mattermost.atlassian.net/browse/PLT-4990
https://mattermost.atlassian.net/browse/PLT-5054

#### Checklist
- [x] Has UI changes
- [x] Includes text changes and localization file updates

### CODE REVIEW

**Comment by grundleborg** [comment]
@esethna this PR is now ready for PM review. Things done:
* System Console permissions are now separate for public/private create/rename/delete.
* UI elements for those features only shown when person has relevant permissions.

Channel Admin permissions level has not been added yet. That's coming in the next PR.

**Comment by esethna** [comment]
Hey George, thanks! Here are some issues I ran into. I was having trouble getting the functionality of the create and delete settings to work. I'm wondering if it has something to do with the sequencing of the server/client PRs? Please take a look:

1. The text dropdown option for the renaming and deletion settings should be `All channel members` instead of `All team members`, since only channel members can access the menu to rename and delete.
image

2. What would be the mana involved in hiding the "+" button and the "Create New Channel" button, rather than just throwing an error if they don't have the appropriate permissions?
image
image

3. Enable public channel creation for: I seem to be able to create PUBLIC channels regardless of the system console setting. I'm also getting a console error when I create a new public channel:

image

4. Enable private group creation for: I'm getting the "You do not have the appropriate permissions" error on the private group creation modal regardless of the system console setting. I think this happened after I changed the setting to "System Admins" and then tried to change it back.

5. FYI I found an issue on master with deleting channels [(see ticket)](https://mattermost.atlassian.net/browse/PLT-5079). 

6. Public channel deletion: Regardless of the system console setting, the "Delete Channel" option doesn't seem to be hidden in the channel menu UI. Delete occurs successfully regardless of permission level.

7. Private group deletion: Regardless of the system console setting, the "Delete Group" option doesn't seem to be hidden in the channel menu UI. I also cannot delete private groups, regardless of the setting, because I get a console error "You do not have the appropriate permissions". I think this happened after I changed the setting to "System Admins" and then tried to change it back.


___
Working as expected:
- Enable public channel renaming for:
- Enable private group renaming for:

**Comment by grundleborg** [comment]
@esethna fixed (1). All the rest should already have been done in the PR. I suspect the problem was the spinmint was using an out-of-date push without the server side changes. Have retriggered the spinmint which hopefully should help.

**Comment by esethna** [comment]
Thanks @grundleborg! I just found one issue with the deletion setting for private groups, it doesn't seem to hide the channel menu option for "Delete group" and delete occurs successfully regardless of your permission level.

Everything else worked great! With the exception of this [known issue for deleting channels](https://mattermost.atlassian.net/browse/PLT-5079)

**Comment by grundleborg** [comment]
@esethna I can't reproduce the issue you described on the spinmint. I tried creating a private group, setting the delete policy to "System Admin", adding a non-system-admin user to it and then trying to delete it as that user. Delete was not in the menu. Is it possible that there was only one member in the group you tested with? We left the feature in place that the last person left in a channel/group can always delete it regardless of their permission level, which would explain that behaviour.

**Comment by esethna** [comment]
@grundleborg yes that was it, thanks for clarifying, that makes sense. Re-tested and verified this works as expected. Thanks!

File: webapp/components/admin_console/policy_settings.jsx [hunk 1]
```diff
@@ -74,46 +82,154 @@ export default class PolicySettings extends AdminSettings {
...
                     label={
                         <FormattedMessage
+                            id='admin.general.policy.restrictPublicChannelCreationTitle'
+                            defaultMessage='Enable public channel creation for:'
+                        />
+                    }
+                    value={this.state.restrictPublicChannelCreation}
+                    onChange={this.handleChange}
+                    helpText={
+                        <FormattedHTMLMessage
```
Comment (enahum) [inline webapp/components/admin_console/policy_settings.jsx]: Why are we using `FormattedHTMLMessage` here?

1. the string does not have any HTML in it
2. remember that we are trying to move away from this component

File: webapp/components/admin_console/policy_settings.jsx [hunk 2]
```diff
@@ -74,46 +82,154 @@ export default class PolicySettings extends AdminSettings {
...
+                            defaultMessage='Enable public channel renaming for:'
                         />
                     }
                     value={this.state.restrictPublicChannelManagement}
                     onChange={this.handleChange}
                     helpText={
                         <FormattedHTMLMessage
                             id='admin.general.policy.restrictPublicChannelManagementDescription'
-                            defaultMessage='Set policy on who can create, delete, rename, and set the header or purpose for public channels.'
+                            defaultMessage='Set policy on who can rename and set the header or purpose for public channels.'
```
Comment (enahum) [inline webapp/components/admin_console/policy_settings.jsx]: same here

File: webapp/components/admin_console/policy_settings.jsx [hunk 3]
```diff
@@ -74,46 +82,154 @@ export default class PolicySettings extends AdminSettings {
...
                     label={
                         <FormattedMessage
+                            id='admin.general.policy.restrictPrivateChannelCreationTitle'
+                            defaultMessage='Enable private group creation for:'
+                        />
+                    }
+                    value={this.state.restrictPrivateChannelCreation}
+                    onChange={this.handleChange}
+                    helpText={
+                        <FormattedHTMLMessage
```
Comment (enahum) [inline webapp/components/admin_console/policy_settings.jsx]: and here

File: webapp/components/admin_console/policy_settings.jsx [hunk 4]
```diff
@@ -74,46 +82,154 @@ export default class PolicySettings extends AdminSettings {
...
+                            defaultMessage='Enable private group renaming for:'
                         />
                     }
                     value={this.state.restrictPrivateChannelManagement}
                     onChange={this.handleChange}
                     helpText={
                         <FormattedHTMLMessage
                             id='admin.general.policy.restrictPrivateChannelManagementDescription'
-                            defaultMessage='Set policy on who can create, delete, rename, and set the header or purpose for private groups.'
+                            defaultMessage='Set policy on who can rename and set the header or purpose for private groups.'
```
Comment (enahum) [inline webapp/components/admin_console/policy_settings.jsx]: also this one

**Comment by grundleborg** [comment]
@enahum good catch. I fixed it where there actually was HTML, but didn't occur to me that there was anything to do where there wasn't any HTML in the message!

**Pull request merged**